### PR TITLE
Assignment4

### PR DESCRIPTION
**Pull request by c-stanton**
Worked with @ColtonKnopik

### CODE REVIEW

File: Logger/entities/Student.cs [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+// Student class inherits the IEntity implementation from Person. So both interface members are inherited implicitly
+// and are publicly accessible as part of the class's API.
+public record class Student : Person
+{ 
+    public string GraduationYear { get; set; }
+
+    public Student(FullName fullName, string graduationYear) : base(fullName)
+    {
+        GraduationYear = graduationYear;
```
Comment (RyanHirte) [inline Logger/entities/Student.cs:11]: May want to consider adding a case where graduationYear is null and how you'd want to handle that. Not really possible to have a student without a year they graduate. Throwing an exception may be good here if it is null.

File: Logger/entities/Student.cs [hunk 2]
```diff
@@ -0,0 +1,21 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
```
Comment (RyanHirte) [inline Logger/entities/Student.cs]: Can be deleted.

File: Logger/entities/Employee.cs [hunk 3]
```diff
@@ -0,0 +1,20 @@
+﻿using System;
```
Comment (RyanHirte) [inline Logger/entities/Employee.cs]: Can be deleted.

File: Logger/entities/Employee.cs [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+// Employee class inherits the IEntity implementation from Person. So both interface members are inherited implicitly
+// and are publicly accessible as part of the class's API.
+public record class Employee : Person
+{
+    public string Position { get; init; }
+    
+    public Employee(FullName fullName, string position) : base(fullName)
+    {
+        Position = position;
```
Comment (RyanHirte) [inline Logger/entities/Employee.cs:11]: Could be good to check for null here, since it's impossible to have an employee that doesn't have a position where they work.

File: Logger/entities/Book.cs [hunk 5]
```diff
@@ -0,0 +1,21 @@
+﻿using System;
```
Comment (RyanHirte) [inline Logger/entities/Book.cs]: Can be deleted.

File: Logger/entities/Book.cs [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+// The Name property is implemented implicitly because it should be publicly accessible as part of the class's API.
+public record class Book : EntityBase
+{
+    public string Title { get; init; }
+    public string Author { get; init; }
+    public override string Name => $"{Title} by {Author}";
+
+    public Book(string title, string author)
+    {
+        Title = title;
```
Comment (RyanHirte) [inline Logger/entities/Book.cs:12]: Making sure either title or author is not null here could be good, as a book needs at least one of those things, that way when you call book.name it returns something that makes sense not " by ".

File: Logger.Tests/StudentTests.cs [hunk 7]
```diff
@@ -0,0 +1,103 @@
+﻿using Xunit;
```
Comment (RyanHirte) [inline Logger.Tests/StudentTests.cs:1]: See comment at student.cs for adding a test checking for null value passed into constructor.

File: Logger.Tests/StorageTests.cs [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+
+        // Assert
+        Assert.False(storage.Contains(entity));
+    }
+
+    [Theory]
+    [InlineData("Book")]
+    [InlineData("Employee")]
+    [InlineData("Student")]
+    public void Get_EntityById_ShouldReturnEntity(string entityType)
```
Comment (RyanHirte) [inline Logger.Tests/StorageTests.cs:73]: Very cool didn't know you could do this.

File: Logger.Tests/StorageTests.cs [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+        Assert.Equal(entity.Name, ((dynamic)retrieved).Name);
+    }
+
+    [Fact]
+    public void Get_NonExistingId_ShouldReturnNull()
+    {
+        // Arrange
+        var storage = new Storage();
+        var book = new Book("1984", "George Orwell");
+        storage.Add(book);
```
Comment (RyanHirte) [inline Logger.Tests/StorageTests.cs]: I don't think adding an object to storage is necessary here since book is never used when it is added to storage.

File: Logger.Tests/StorageTests.cs [hunk 10]
```diff
@@ -0,0 +1,128 @@
+﻿using Logger.entities;
```
Comment (RyanHirte) [inline Logger.Tests/StorageTests.cs:1]: Could add a test seeing what happens if an entity that doesn't exist in storage is removed.

File: Logger.Tests/EmployeeTests.cs [hunk 11]
```diff
@@ -0,0 +1,99 @@
+﻿using Xunit;
```
Comment (RyanHirte) [inline Logger.Tests/EmployeeTests.cs:1]: See Employee.cs for adding test for null position passed into constructor.

File: Logger.Tests/BookTests.cs [hunk 12]
```diff
@@ -0,0 +1,77 @@
+﻿using Xunit;
```
Comment (RyanHirte) [inline Logger.Tests/BookTests.cs:1]: See Book.cs for checking null author or title.

**Review by RyanHirte — COMMENTED** [review]
Define a full name record (first, last, middle) handling optional and null appropriately. ✔
Provide a comment on the full name record on why you selected to define a value or a reference type and ✔
Provide a comment on the full name record on why or why not the type is immutable. ✔
Define an IEntity interface: ✔
Add an Id property of type Guid that is an init-only setter. ✔
Add a Name property that is string. ✔
Define an abstract base class that implements IEntity - appropriately choosing to implement the interface explicitly or implicitly. ✔
Do not implement the Name property in this abstract class. ✔
Do force any derived classes to provide an implementation for Name. ✔
Define book, student, and employee records - all with entity capabilities. ✔
Choose a thoughtful approach for the value of Name for each type of entity. In all cases, there should be no backing field for Name (not even one generated by an automatically implemented property) - consider a calculated property. ✔
Provide a comment on each interface member in each entity explaining why you implemented it implicitly or explicitly. ✔
You should consider the relationship between Student and Employee and refactor the common code shared between them. ✔
Given the properties on the full name record, you should consider which entities it makes sense to use it. ✔
Test that the equality behavior on these entities behaves as expected. ✔
Provide a comment on each interface method explaining why you implemented it implicitly or explicitly.
Unit test the Storage class using various entities to ensure that there are no bugs.
Code review (using TODO: comments if there are any issues - which there may not be) all logger-related classes.)
Pay special attention to the CreateLogger methods throughout.
Fundamentals
Be sure you enable:
Enable Nullability for all projects ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Ensure that you turn on EnforceCodeStyleInBuild ✔
Ensure that you turn on Nullability (Nullable) ✔
Ensure there are no errors or warnings (including code analysis warnings) ✔
All of the above should be unit tested ✔
Choose simplicity over complexity ✔

Overall great job!

File: Logger/entities/Book.cs (hunk 6 above)
Comment (c-stanton) [reply Logger/entities/Book.cs:12]: Thanks! I added your suggestion.

File: Logger/entities/Student.cs (hunk 1 above)
Comment (c-stanton) [reply Logger/entities/Student.cs:11]: Thank you! I added your suggestion.

File: Logger/entities/Employee.cs (hunk 4 above)
Comment (c-stanton) [reply Logger/entities/Employee.cs:11]: I added your suggestion, thank you!

**Comment by PigPig-SDK** [comment]
Define a full name record (first, last, middle) handling optional and null appropriately. ✔
Provide a comment on the full name record on why you selected to define a value or a reference type and ✔
Provide a comment on the full name record on why or why not the type is immutable. ✔
Define an IEntity interface: ✔
Add an Id property of type Guid that is an init-only setter. ✔
Add a Name property that is string. ✔
Define an abstract base class that implements IEntity - appropriately choosing to implement the interface explicitly or implicitly. ✔
Do not implement the Name property in this abstract class. ✔
Do force any derived classes to provide an implementation for Name. ✔
Define book, student, and employee records - all with entity capabilities. ✔
Choose a thoughtful approach for the value of Name for each type of entity. In all cases, there should be no backing field for Name (not even one generated by an automatically implemented property) - consider a calculated property. ✔
Provide a comment on each interface member in each entity explaining why you implemented it implicitly or explicitly. ✔
You should consider the relationship between Student and Employee and refactor the common code shared between them. ✔
Given the properties on the full name record, you should consider which entities it makes sense to use it. ✔
Test that the equality behavior on these entities behaves as expected. ✔
Provide a comment on each interface method explaining why you implemented it implicitly or explicitly.
Unit test the Storage class using various entities to ensure that there are no bugs.
Code review (using TODO: comments if there are any issues - which there may not be) all logger-related classes.)
Pay special attention to the CreateLogger methods throughout.
Fundamentals
Be sure you enable:
Enable Nullability for all projects ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Ensure that you turn on EnforceCodeStyleInBuild ✔
Ensure that you turn on Nullability (Nullable) ✔
Ensure there are no errors or warnings (including code analysis warnings) ✔
All of the above should be unit tested ✔
Choose simplicity over complexity ✔

File: Logger.Tests/StorageTests.cs (hunk 9 above)
Comment (ColtonKnopik) [reply Logger.Tests/StorageTests.cs]: Yes that is a good point it has been removed.

File: Logger.Tests/StorageTests.cs (hunk 10 above)
Comment (ColtonKnopik) [reply Logger.Tests/StorageTests.cs:1]: nice, added

**Comment by JosephPotapenko** [comment]
Define a full name record (first, last, middle) handling optional and null appropriately. ✔
Provide a comment on the full name record on why you selected to define a value or a reference type and ✔
Provide a comment on the full name record on why or why not the type is immutable. ✔
Define an IEntity interface: ✔
Add an Id property of type Guid that is an init-only setter. ✔
Add a Name property that is string. ✔
Define an abstract base class that implements IEntity - appropriately choosing to implement the interface explicitly or implicitly. ✔
Do not implement the Name property in this abstract class. ✔
Do force any derived classes to provide an implementation for Name. ✔
Define book, student, and employee records - all with entity capabilities. ✔
Choose a thoughtful approach for the value of Name for each type of entity. In all cases, there should be no backing field for Name (not even one generated by an automatically implemented property) - consider a calculated property. ✔
Provide a comment on each interface member in each entity explaining why you implemented it implicitly or explicitly. ✔
You should consider the relationship between Student and Employee and refactor the common code shared between them. ✔
Given the properties on the full name record, you should consider which entities it makes sense to use it. ✔
Test that the equality behavior on these entities behaves as expected. ✔
Provide a comment on each interface method explaining why you implemented it implicitly or explicitly.
Unit test the Storage class using various entities to ensure that there are no bugs. ✔
Code review (using TODO: comments if there are any issues - which there may not be) all logger-related classes.) ✔
Pay special attention to the CreateLogger methods throughout.
Fundamentals
Be sure you enable:
Enable Nullability for all projects ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Ensure that you turn on EnforceCodeStyleInBuild ✔
Ensure that you turn on Nullability (Nullable) ✔
Ensure there are no errors or warnings (including code analysis warnings) ✔
All of the above should be unit tested ✔
Choose simplicity over complexity ✔

Apart from what was mentioned above, it looks really good! Not much I could find to nitpick about.

File: Logger/FullName.cs [hunk 13]
```diff
@@ -0,0 +1,20 @@
+﻿namespace Logger;
+
+// The FullName is defined as a record class to leverage records built in value based equality and immutability.
+// Which is valuable for representing data objects like names where the focus is on the data itself rather than identity.
+// The record is immutable because all of its properties have init-only setters.
+public record class FullName
+{
+    public string First { get; init; }
+    public string? Middle { get; init; }
```
Comment (BillMillerCoding) [inline Logger/FullName.cs:9]: It seems to me that this string doesn't need to be marked nullable. It is always prevented from being null in the constructor. If there is something I am not seeing please lmk :)

File: Logger/Person.cs [hunk 14]
```diff
@@ -0,0 +1,14 @@
+﻿namespace Logger;
+
+public abstract record Person : EntityBase
+{
+    public FullName FullName { get; init; }
+
+    // This is implemented explicitly because it is clearer and more consistent with the IEntity contract that Name is a string 
+    public override string Name => FullName.ToString();
```
Comment (BillMillerCoding) [inline Logger/Person.cs:8]: I don't think this counts as explicit. I think it must be written IEntity.Name if it is to be technically explicit. But since in this case you are mixing Interfaces with inheritance I could be wrong. But since you dont call it explicitly in BaseEntity, I would imagine it doesnt count as explicit.

**Review by BillMillerCoding — COMMENTED** [review]
Define a full name record (first, last, middle) handling optional and null appropriately. ✔
Provide a comment on the full name record on why you selected to define a value or a reference type and ✔
Provide a comment on the full name record on why or why not the type is immutable. ✔
Define an IEntity interface: ✔
Add an Id property of type Guid that is an init-only setter. ✔
Add a Name property that is string. ✔
Define an abstract base class that implements IEntity - appropriately choosing to implement the interface explicitly or implicitly. ✔
Do not implement the Name property in this abstract class. ✔
Do force any derived classes to provide an implementation for Name. ✔
Define book, student, and employee records - all with entity capabilities. ✔
Choose a thoughtful approach for the value of Name for each type of entity. In all cases, there should be no backing field for Name (not even one generated by an automatically implemented property) - consider a calculated property. ✔
Provide a comment on each interface member in each entity explaining why you implemented it implicitly or explicitly. ✔
You should consider the relationship between Student and Employee and refactor the common code shared between them. ✔
Given the properties on the full name record, you should consider which entities it makes sense to use it. ✔
Test that the equality behavior on these entities behaves as expected. ✔
Provide a comment on each interface method explaining why you implemented it implicitly or explicitly.✔
Unit test the Storage class using various entities to ensure that there are no bugs.✔
Code review (using TODO: comments if there are any issues - which there may not be) all logger-related classes.) X
Pay special attention to the CreateLogger methods throughout.
Fundamentals
Be sure you enable:
Enable Nullability for all projects ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Ensure that you turn on EnforceCodeStyleInBuild ✔
Ensure that you turn on Nullability (Nullable) ✔
Ensure there are no errors or warnings (including code analysis warnings) ✔
All of the above should be unit tested ✔
Choose simplicity over complexity ✔

File: Logger/entities/Student.cs (hunk 2 above)
Comment (c-stanton) [reply Logger/entities/Student.cs]: Thanks! My partner did clean up

File: Logger/entities/Employee.cs (hunk 3 above)
Comment (c-stanton) [reply Logger/entities/Employee.cs]: Cleanup was completed

File: Logger/entities/Book.cs (hunk 5 above)
Comment (c-stanton) [reply Logger/entities/Book.cs]: Also removed, thank you.

File: Logger.Tests/StudentTests.cs (hunk 7 above)
Comment (c-stanton) [reply Logger.Tests/StudentTests.cs:1]: My partner added the test, thanks!

File: Logger.Tests/EmployeeTests.cs (hunk 11 above)
Comment (c-stanton) [reply Logger.Tests/EmployeeTests.cs:1]: My partner updated it, thank you!

File: Logger.Tests/BookTests.cs (hunk 12 above)
Comment (c-stanton) [reply Logger.Tests/BookTests.cs:1]: My partner added it, thanks

File: Logger/Person.cs (hunk 14 above)
Comment (c-stanton) [reply Logger/Person.cs:8]: You are correct, it is not explicit interface implementation but rather implicit. I will update the comment to reflect that fact, nice catch!

File: Logger/FullName.cs (hunk 13 above)
Comment (c-stanton) [reply Logger/FullName.cs:9]: You're correct, the null middle name was always going to be converted to and treated like an empty string making null and empty string functionally equal. I updated the code to allow the constructor to a store a middle name as null. Thank you!

File: Logger/FullName.cs [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+    public string Last { get; init; }
+    public FullName(string first, string? middle, string last)
+    {
+        First = first;
+        Middle = middle;
+        Last = last;
+    }
```
Comment (Joshua-Lester3) [inline Logger/FullName.cs:16]: I'd add some null handling for first and last params

File: Logger/FullName.cs [hunk 16]
```diff
@@ -0,0 +1,20 @@
+﻿namespace Logger;
+
+// The FullName is defined as a record class to leverage records built in value based equality and immutability.
+// Which is valuable for representing data objects like names where the focus is on the data itself rather than identity.
+// The record is immutable because all of its properties have init-only setters.
```
Comment (Joshua-Lester3) [inline Logger/FullName.cs:5]: I'll give points for the immutability reasoning requirement, but I think the requirement was specifying why you made it immutable vs mutable, not how it is.

File: Logger/entities/Book.cs [hunk 17]
```diff
@@ -0,0 +1,25 @@
+﻿namespace Logger.entities;
```
Comment (Joshua-Lester3) [inline Logger/entities/Book.cs:1]: There's a guideline to make namespace paths/folders PascalCase: https://essentialcsharp.com/guidelines#naming-conventions-table

**Review by Joshua-Lester3 — COMMENTED** [review]
## Tasks

- Define a full name record (first, last, middle) handling optional and null appropriately. ❌ Null not appropriately handled - no null checks
  - Provide a comment on the full name record on why you selected to define a value or a reference type and ❌ No comment on why it's value or reference
  - Provide a comment on the full name record on why or why not the type is immutable. ✔
- Define an `IEntity` interface: ✔
  - Add an Id property of type `Guid` that is an init-only setter. ✔
  - Add a Name property that is `string`. ✔
- Define an abstract base class that implements `IEntity` - appropriately choosing to implement the interface explicitly or implicitly. ✔
  - Do not implement the `Name` property in this abstract class. ✔
  - Do force any derived classes to provide an implementation for `Name`. ✔
- Define book, student, and employee records - all with entity capabilities. ✔
  - Choose a thoughtful approach for the value of `Name` for each type of entity. In all cases, there should be no backing field for Name (not even one generated by an automatically implemented property) - consider a calculated property. ✔
  - Provide a comment on each interface **member** in each entity explaining why you implemented it implicitly or explicitly. ✔
  - You should consider the relationship between Student and Employee and refactor the common code shared between them. ✔
  - Given the properties on the full name record, you should consider which entities it makes sense to use it. ✔
  - Test that the equality behavior on these entities behaves as expected. ✔
- Unit test the Storage class using various entities to ensure that there are no bugs. ✔
- Code review (using TODO: comments if there are any issues - which there may not be) all logger-related classes.) ✔
  - Pay special attention to the CreateLogger methods throughout. ✔

## Fundamentals

- Be sure you enable:
  - Enable Nullability for all projects ✔
  - Set `LangVersion` and the `TargetFramework` to the latest released versions available (preview versions optional) ✔
  - Ensure that you turn on code analysis for all projects(`EnableNETAnalyzers`) ✔
  - Ensure that you turn on `EnforceCodeStyleInBuild` ✔
  - Ensure that you turn on Nullability (`Nullable`) ✔
- **Ensure there are no errors or warnings (including code analysis warnings)** ❌ Not added
- **All of the above should be unit tested** ✔
- **Choose simplicity over complexity** ✔

## Extra Credit

- Refactor your project files to put common properties and/or items into a single file. ❌
- Convert `ILogger.CreateLogger` to a generic method. ❌
  - Returns the logger classes type that is to be created. ❌
  - Has a type parameter for the configuration type ❌
